File: server/lib/routes/categories.dart
import 'dart:convert';
import 'package:shelf/shelf.dart';
import 'package:shelf_router/shelf_router.dart';
import '../store/json_store.dart';
import '../services/cleaning_service.dart';

Router categoryRoutes() {
  final router = Router();

  // GET /api/categories - List all user-defined category rules
  router.get('/api/categories', (Request request) {
    final rules = loadCategoryRules();
    return Response.ok(
      jsonEncode(rules.map((r) => r.toJson()).toList()),
      headers: {'Content-Type': 'application/json'},
    );
  });

  // GET /api/categories/should-import - Check if defaults should be imported
  router.get('/api/categories/should-import', (Request request) {
    final rules = loadCategoryRules();
    final shouldImport = rules.isEmpty;
    return Response.ok(
      jsonEncode({'shouldImport': shouldImport, 'count': rules.length}),
      headers: {'Content-Type': 'application/json'},
    );
  });

  // POST /api/categories/import-defaults - Import default rules to user rules
  router.post('/api/categories/import-defaults', (Request request) async {
    try {
      final defaultRules = loadDefaultCategoryRules();

      if (defaultRules.isEmpty) {
        return Response.badRequest(
          body: jsonEncode({'error': 'No default rules available to import'}),
          headers: {'Content-Type': 'application/json'},
        );
      }

      // Save default rules as user rules
      saveCategoryRules(defaultRules);

      // Trigger re-categorization
      cleanAllTransactions();

      return Response.ok(
        jsonEncode({
          'success': true,
          'imported': defaultRules.length,
          'message': 'Imported ${defaultRules.length} default rules'
        }),
        headers: {'Content-Type': 'application/json'},
      );
    } catch (e) {
      return Response.internalServerError(
        body: jsonEncode({'error': e.toString()}),
        headers: {'Content-Type': 'application/json'},
      );
    }
  });

  // POST /api/categories - Add new regex pattern
  router.post('/api/categories', (Request request) async {
    try {
      final body = await request.readAsString();
      final json = jsonDecode(body) as Map<String, dynamic>;

      final pattern = json['pattern'] as String?;
      final category = json['category'] as String?;
      final caseSensitive = json['caseSensitive'] as bool? ?? false;

      if (pattern == null || pattern.isEmpty) {
        return Response.badRequest(
          body: jsonEncode({'error': 'Pattern is required'}),
          headers: {'Content-Type': 'application/json'},
        );
      }

      if (category == null || category.isEmpty) {
        return Response.badRequest(
          body: jsonEncode({'error': 'Category is required'}),
          headers: {'Content-Type': 'application/json'},
        );
      }

      // Validate regex pattern
      try {
        RegExp(pattern, caseSensitive: caseSensitive);
      } catch (e) {
        return Response.badRequest(
          body: jsonEncode({'error': 'Invalid regex pattern: $e'}),
          headers: {'Content-Type': 'application/json'},
        );
      }

      final rules = loadCategoryRules();
      final newRule = CategoryRule(
        id: DateTime.now().millisecondsSinceEpoch.toString(),
        category: category,
        pattern: pattern,
        caseSensitive: caseSensitive,
      );

      rules.add(newRule);
      saveCategoryRules(rules);

      // Trigger re-categorization
      cleanAllTransactions();

      return Response.ok(
        jsonEncode(newRule.toJson()),
        headers: {'Content-Type': 'application/json'},
      );
    } catch (e) {
      return Response.internalServerError(
        body: jsonEncode({'error': e.toString()}),
        headers: {'Content-Type': 'application/json'},
      );
    }
  });

  // PUT /api/categories/<id> - Update regex pattern
  router.put('/api/categories/<id>', (Request request, String id) async {
    try {
      final body = await request.readAsString();
      final json = jsonDecode(body) as Map<String, dynamic>;

      final rules = loadCategoryRules();
      final index = rules.indexWhere((r) => r.id == id);

      if (index == -1) {
        return Response.notFound(
          jsonEncode({'error': 'Category rule not found'}),
          headers: {'Content-Type': 'application/json'},
        );
      }

      final pattern = json['pattern'] as String?;
      final category = json['category'] as String?;
      final caseSensitive = json['caseSensitive'] as bool?;

      if (pattern != null && pattern.isEmpty) {
        return Response.badRequest(
          body: jsonEncode({'error': 'Pattern cannot be empty'}),
          headers: {'Content-Type': 'application/json'},
        );
      }

      if (category != null && category.isEmpty) {
        return Response.badRequest(
          body: jsonEncode({'error': 'Category cannot be empty'}),
          headers: {'Content-Type': 'application/json'},
        );
      }

      // Validate regex pattern if provided
      if (pattern != null) {
        try {
          RegExp(pattern, caseSensitive: caseSensitive ?? rules[index].caseSensitive);
        } catch (e) {
          return Response.badRequest(
            body: jsonEncode({'error': 'Invalid regex pattern: $e'}),
            headers: {'Content-Type': 'application/json'},
          );
        }
      }

      final updatedRule = CategoryRule(
        id: id,
        category: category ?? rules[index].category,
        pattern: pattern ?? rules[index].pattern,
        caseSensitive: caseSensitive ?? rules[index].caseSensitive,
      );

      rules[index] = updatedRule;
      saveCategoryRules(rules);

      // Trigger re-categorization
      cleanAllTransactions();

      return Response.ok(
        jsonEncode(updatedRule.toJson()),
        headers: {'Content-Type': 'application/json'},
      );
    } catch (e) {
      return Response.internalServerError(
        body: jsonEncode({'error': e.toString()}),
        headers: {'Content-Type': 'application/json'},
      );
    }
  });

  // DELETE /api/categories/<id> - Delete regex pattern
  router.delete('/api/categories/<id>', (Request request, String id) {
    try {
      final rules = loadCategoryRules();
      final initialLength = rules.length;
      rules.removeWhere((r) => r.id == id);

      if (rules.length == initialLength) {
        return Response.notFound(
          jsonEncode({'error': 'Category rule not found'}),
          headers: {'Content-Type': 'application/json'},
        );
      }

      saveCategoryRules(rules);

      // Trigger re-categorization
      cleanAllTransactions();

      return Response.ok(
        jsonEncode({'success': true}),
        headers: {'Content-Type': 'application/json'},
      );
    } catch (e) {
      return Response.internalServerError(
        body: jsonEncode({'error': e.toString()}),
        headers: {'Content-Type': 'application/json'},
      );
    }
  });

  // GET /api/transactions/uncategorized - Get all uncategorized transactions
  router.get('/api/transactions/uncategorized', (Request request) {
    final transactions = loadCleanTransactions();
    final uncategorized = transactions
        .where((tx) => tx.data['category'] == 'Uncategorized')
        .toList();

    return Response.ok(
      jsonEncode(uncategorized.map((t) => t.toJson()).toList()),
      headers: {'Content-Type': 'application/json'},
    );
  });

  // POST /api/transactions/<id>/categorize - Manually categorize a transaction
  router.post('/api/transactions/<id>/categorize',
      (Request request, String id) async {
    try {
      final body = await request.readAsString();
      final json = jsonDecode(body) as Map<String, dynamic>;

      final category = json['category'] as String?;
      final createRule = json['createRule'] as bool? ?? false;
      final rulePattern = json['rulePattern'] as String?;

      if (category == null || category.isEmpty) {
        return Response.badRequest(
          body: jsonEncode({'error': 'Category is required'}),
          headers: {'Content-Type': 'application/json'},
        );
      }

      // Validate regex first before making any changes
      if (createRule && rulePattern != null && rulePattern.isNotEmpty) {
        try {
          RegExp(rulePattern, caseSensitive: false);
        } catch (e) {
          return Response.badRequest(
            body: jsonEncode({'error': 'Invalid regex pattern: $e'}),
            headers: {'Content-Type': 'application/json'},
          );
        }
      }

      // Load overrides and add this transaction
      final overrides = loadOverrides();
      overrides[id] = category;
      saveOverrides(overrides);

      // Create the rule (already validated above)
      if (createRule && rulePattern != null && rulePattern.isNotEmpty) {
        final rules = loadCategoryRules();
        final newRule = CategoryRule(
          id: DateTime.now().millisecondsSinceEpoch.toString(),
          category: category,
          pattern: rulePattern,
          caseSensitive: false,
        );
        rules.add(newRule);
        saveCategoryRules(rules);
      }

      // Trigger re-categorization
      cleanAllTransactions();

      return Response.ok(
        jsonEncode({'success': true, 'category': category}),
        headers: {'Content-Type': 'application/json'},
      );
    } catch (e) {
      return Response.internalServerError(
        body: jsonEncode({'error': e.toString()}),
        headers: {'Content-Type': 'application/json'},
      );
    }
  });

  return router;
}

File: server/src/routes/categories.ts
import { Hono } from 'hono';
import { cleanAllTransactions } from '../services/cleaningService.js';
import {
  loadCategoryRules,
  loadCleanTransactions,
  loadDefaultCategoryRules,
  loadOverrides,
  saveCategoryRules,
  saveOverrides,
} from '../store/jsonStore.js';
import type { CategoryRule } from '../types.js';

const app = new Hono();

app.get('/api/categories', (c) => {
  return c.json(loadCategoryRules());
});

app.post('/api/categories', async (c) => {
  try {
    const body = await c.req.json<{
      pattern?: string;
      category?: string;
      caseSensitive?: boolean;
    }>();
    const { pattern, category, caseSensitive = false } = body;

    if (!pattern) return c.json({ error: 'Pattern is required' }, 400);
    if (!category) return c.json({ error: 'Category is required' }, 400);

    try {
      new RegExp(pattern, caseSensitive ? '' : 'i');
    } catch (e) {
      return c.json({ error: `Invalid regex pattern: ${e}` }, 400);
    }

    const rules = loadCategoryRules();
    const newRule: CategoryRule = {
      id: String(Date.now()),
      category,
      pattern,
      caseSensitive,
    };
    rules.push(newRule);
    saveCategoryRules(rules);

    return c.json(newRule);
  } catch (err) {
    return c.json({ error: String(err) }, 500);
  }
});

app.get('/api/categories/should-import-defaults', (c) => {
  const rules = loadCategoryRules();
  return c.json({ shouldImport: rules.length === 0, count: rules.length });
});

app.post('/api/categories/import-defaults', async (c) => {
  try {
    const defaultRules = loadDefaultCategoryRules();
    if (defaultRules.length === 0) {
      return c.json({ error: 'No default rules available to import' }, 400);
    }
    saveCategoryRules(defaultRules);
    cleanAllTransactions();
    return c.json({
      success: true,
      imported: defaultRules.length,
      message: `Imported ${defaultRules.length} default rules`,
    });
  } catch (err) {
    return c.json({ error: String(err) }, 500);
  }
});

app.put('/api/categories/:id', async (c) => {
  try {
    const id = c.req.param('id');
    const body = await c.req.json<{
      pattern?: string;
      category?: string;
      caseSensitive?: boolean;
    }>();

    const rules = loadCategoryRules();
    const index = rules.findIndex((r) => r.id === id);
    if (index === -1) return c.json({ error: 'Category rule not found' }, 404);

    const updatedRule: CategoryRule = {
      id,
      category: body.category ?? rules[index].category,
      pattern: body.pattern ?? rules[index].pattern,
      caseSensitive: body.caseSensitive ?? rules[index].caseSensitive,
    };
    rules[index] = updatedRule;
    saveCategoryRules(rules);
    cleanAllTransactions();

    return c.json(updatedRule);
  } catch (err) {
    return c.json({ error: String(err) }, 500);
  }
});

app.delete('/api/categories/:id', (c) => {
  try {
    const id = c.req.param('id');
    const rules = loadCategoryRules();
    const initialLength = rules.length;
    const filtered = rules.filter((r) => r.id !== id);
    if (filtered.length === initialLength) {
      return c.json({ error: 'Category rule not found' }, 404);
    }
    saveCategoryRules(filtered);
    cleanAllTransactions();
    return c.json({ success: true });
  } catch (err) {
    return c.json({ error: String(err) }, 500);
  }
});

app.post('/api/transactions/:id/categorize', async (c) => {
  try {
    const id = c.req.param('id');
    const body = await c.req.json<{
      category?: string;
      createRule?: boolean;
      rulePattern?: string;
    }>();
    const { category, createRule = false, rulePattern } = body;

    if (!category) return c.json({ error: 'Category is required' }, 400);

    const overrides = loadOverrides();
    overrides[id] = category;
    saveOverrides(overrides);

    if (createRule && rulePattern) {
      try {
        new RegExp(rulePattern, 'i');
        const rules = loadCategoryRules();
        rules.push({
          id: String(Date.now()),
          category,
          pattern: rulePattern,
          caseSensitive: false,
        });
        saveCategoryRules(rules);
      } catch (e) {
        return c.json({ error: `Invalid regex pattern: ${e}` }, 400);
      }
    }

    cleanAllTransactions();
    return c.json({ success: true, category });
  } catch (err) {
    return c.json({ error: String(err) }, 500);
  }
});

app.get('/api/transactions/uncategorized', (c) => {
  const transactions = loadCleanTransactions();
  return c.json(transactions.filter((tx) => tx.category === 'Uncategorized'));
});

export default app;
